### PR TITLE
[otlp] Use LogRecord.Logger.Name for instrumentation scope if LogRecord.CategoryName is null

### PR DESCRIPTION
**Pull request by AB027PS**
Fixes https://github.com/open-telemetry/opentelemetry-dotnet/issues/5274

## Changes

I made a change to OtlpLogRecordTransformer to address a situation where LogRecord.CategoryName is null
according to a suggestion in https://github.com/open-telemetry/opentelemetry-dotnet/issues/5274

## Merge requirement checklist

* [ ] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [ ] Unit tests added/updated
* [ ] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [ ] Changes in public API reviewed (if applicable)

### CODE REVIEW

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs [hunk 1]
```diff
@@ -16,6 +16,8 @@ namespace OpenTelemetry.Exporter.OpenTelemetryProtocol.Implementation;
 
 internal sealed class OtlpLogRecordTransformer
 {
+    private const string DefaultScopeName = "unknown_scope";
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:19]: Is this part of the spec. Why cannot we put logs without logger name individually under a new scope?

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs [hunk 2]
```diff
@@ -47,10 +49,11 @@ public OtlpLogRecordTransformer(SdkLimitOptions sdkLimitOptions, ExperimentalOpt
             var otlpLogRecord = this.ToOtlpLog(logRecord);
             if (otlpLogRecord != null)
             {
-                if (!logsByCategory.TryGetValue(logRecord.CategoryName, out var scopeLogs))
+                var scopeName = logRecord.CategoryName ?? logRecord.Logger?.Name ?? DefaultScopeName;
```
Comment (vishweshbankwar) [inline src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:52]: Could we just update the sdk to set the logger name to categoryname in `OpenTelemetryLogger`  and use just the logger name here? cc: @CodeBlanch

**Comment by CodeBlanch** [comment]
@AB027PS This is good thanks! There are a few things we need to do before we can merge:

* Resolve warnings. Looks like some style issues. If you are using VS switch to `Release` config and rebuild that should surface everything for you locally.

* Update CHANGELOG. Add a note at the bottom of the unreleased section here: https://github.com/open-telemetry/opentelemetry-dotnet/blob/8b54074c7a392b0e295fa7aa0f1a3ab1cb28025a/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/CHANGELOG.md?plain=1#L23

   Something like this...

   ```
   * Updated `OtlpLogExporter` to set instrumentation scope name on the data model
     from `LogRecord.Logger.Name` if `LogRecord.CategoryName` is `null`. This is
     typically the case when using the experimental Logs Bridge API.
     ([#5300](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5300))
   ```

* We need a unit test. In here: https://github.com/open-telemetry/opentelemetry-dotnet/blob/8b54074c7a392b0e295fa7aa0f1a3ab1cb28025a/test/OpenTelemetry.Exporter.OpenTelemetryProtocol.Tests/OtlpLogExporterTests.cs#L648

   Drop a test like...

   ```csharp
    [Fact]
    public void LogRecordLoggerNameIsExportedWhenUsingBridgeApi()
    {
        LogRecordAttributeList attributes = default;
        attributes.Add("name", "tomato");
        attributes.Add("price", 2.99);
        attributes.Add("{OriginalFormat}", "Hello from {name} {price}.");

        var logRecords = new List<LogRecord>();

        using (var loggerProvider = Sdk.CreateLoggerProviderBuilder()
            .AddInMemoryExporter(logRecords)
            .Build())
        {
            var logger = loggerProvider.GetLogger("my_instrumentation_scope_name");

            logger.EmitLog(new LogRecordData(), attributes);
        }

        Assert.Single(logRecords);

        var otlpLogRecordTransformer = new OtlpLogRecordTransformer(DefaultSdkLimitOptions, new());

        var batch = new Batch<LogRecord>(new[] { logRecords[0] }, 1);

        var request = otlpLogRecordTransformer.BuildExportRequest(
            new Proto.Resource.V1.Resource(),
            batch);

        Assert.NotNull(request);
        Assert.Single(request.ResourceLogs);
        Assert.Single(request.ResourceLogs[0].ScopeLogs);

        Assert.Equal("my_instrumentation_scope_name", request.ResourceLogs[0].ScopeLogs[0].Scope?.Name);
    }
   ```

If you need help with anything let me know, happy to assist.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs (hunk 2 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:52]: We could. Right now for ILogger path we set [a Logger](https://github.com/open-telemetry/opentelemetry-dotnet/blob/8b54074c7a392b0e295fa7aa0f1a3ab1cb28025a/src/OpenTelemetry/Logs/ILogger/OpenTelemetryLogger.cs#L100) which is named [OpenTelemetry w/ sdk version](https://github.com/open-telemetry/opentelemetry-dotnet/blob/8b54074c7a392b0e295fa7aa0f1a3ab1cb28025a/src/OpenTelemetry/Logs/ILogger/OpenTelemetryLogger.cs#L252). But we could modify the [ctor](https://github.com/open-telemetry/opentelemetry-dotnet/blob/8b54074c7a392b0e295fa7aa0f1a3ab1cb28025a/src/OpenTelemetry/Logs/ILogger/OpenTelemetryLogger.cs#L40) to build a Logger from categoryName.

My only concern would be what to do if `LogRecord.CategoryName` is changed (like inside a processor). It [has a public setter](https://github.com/open-telemetry/opentelemetry-dotnet/blob/8b54074c7a392b0e295fa7aa0f1a3ab1cb28025a/src/OpenTelemetry/Logs/LogRecord.cs#L165). Would we generate a new Logger? Could be an unexpected perf hit to do that.

File: src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs (hunk 1 above)
Comment (CodeBlanch) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:19]: Probably a more sensible default is just `string.Empty`. If you request a logger without supplying a name (eg `loggerProvider.GetLogger()`), that's [what you'll get](https://github.com/open-telemetry/opentelemetry-dotnet/blob/8b54074c7a392b0e295fa7aa0f1a3ab1cb28025a/src/OpenTelemetry.Api/Logs/Logger.cs#L37). In practice we should never run into this default case as there will always be a `Logger` either via ILogger path or Log Bridge API. But good to be defensive here I think.

Comment (vishweshbankwar) [reply src/OpenTelemetry.Exporter.OpenTelemetryProtocol/Implementation/OtlpLogRecordTransformer.cs:19]: Empty sounds reasonable to me.

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Pull request merged**